diff --git a/source/tinycthread.h b/source/threads.h
similarity index 97%
rename from source/tinycthread.h
rename to source/threads.h
index 2fc7f62..c42b430 100644
--- a/source/tinycthread.h
+++ b/source/threads.h
@@ -160,6 +160,16 @@ int _tthread_timespec_get(struct timespec *ts, int base);
  #define _Thread_local __thread
 #endif
 
+#if defined __GNUC__ && defined __GNUC_MINOR__
+ #define __GNUC_PREREQ(maj, min) ((__GNUC__ << 16) + __GNUC_MINOR__ >= ((maj) << 16) + (min))
+#else
+ #define __GNUC_PREREQ(maj, min) 0
+#endif
+
+#if (!defined __STDC_VERSION__ || __STDC_VERSION__ <= 201710L || !__GNUC_PREREQ (13, 0)) && !defined __cplusplus
+ #define thread_local _Thread_local
+#endif
+
 /* Macros */
 #if defined(_TTHREAD_WIN32_)
 #define TSS_DTOR_ITERATIONS (4)
diff --git a/source/tinycthread.c b/source/tinycthread.c
index 195e413..8d7efb6 100644
--- a/source/tinycthread.c
+++ b/source/tinycthread.c
@@ -22,7 +22,7 @@ freely, subject to the following restrictions:
     distribution.
 */
 
-#include "tinycthread.h"
+#include "threads.h"
 #include <stdlib.h>
 
 /* Platform specific includes */
@@ -152,7 +152,7 @@ int mtx_timedlock(mtx_t *mtx, const struct timespec *ts)
   else
   {
     timeoutMs  = (DWORD)(ts->tv_sec  - current_ts.tv_sec)  * 1000;
-    timeoutMs += (ts->tv_nsec - current_ts.tv_nsec) / 1000000;
+    timeoutMs += (DWORD)((ts->tv_nsec - current_ts.tv_nsec) / 1000000);
     timeoutMs += 1;
   }
 
@@ -452,8 +452,8 @@ int cnd_timedwait(cnd_t *cond, mtx_t *mtx, const struct timespec *ts)
   struct timespec now;
   if (timespec_get(&now, TIME_UTC) == TIME_UTC)
   {
-    unsigned long long nowInMilliseconds = now.tv_sec * 1000 + now.tv_nsec / 1000000;
-    unsigned long long tsInMilliseconds  = ts->tv_sec * 1000 + ts->tv_nsec / 1000000;
+    unsigned long long nowInMilliseconds = (unsigned long long)now.tv_sec * 1000 + (unsigned long long)now.tv_nsec / 1000000;
+    unsigned long long tsInMilliseconds  = (unsigned long long)ts->tv_sec * 1000 + (unsigned long long)ts->tv_nsec / 1000000;
     DWORD delta = (tsInMilliseconds > nowInMilliseconds) ?
       (DWORD)(tsInMilliseconds - nowInMilliseconds) : 0;
     return _cnd_timedwait_win32(cond, mtx, delta);
@@ -921,6 +921,8 @@ void call_once(once_flag *flag, void (*func)(void))
         EnterCriticalSection(&(flag->lock));
         LeaveCriticalSection(&(flag->lock));
         break;
+      default:
+        break;
     }
   }
 }
